redis/src/tls.rs

```diff
@@ -0,0 +1,174 @@
+///     .expect("Unable to handle client key");
+/// let ca_file = File::open(ca_file).expect("Cannot open CA cert file");
+/// let mut reader = BufReader::new(ca_file);
+/// tls_builder
+///     .root_cert(&mut reader)
+///     .expect("Unable to load CA cert");
+///
+/// let client = tls_builder.build().expect("Unable to build client");
+/// ```
+///
+impl TLSClientBuilder {
+    /// Initiate builder with given _server_ and _port_
+    pub fn new(host: String, port: u16) -> TLSClientBuilder {
+        TLSClientBuilder {
+            host,
+            port,
+            client_key: None,
+            client_certs: Vec::new(),
+            root_cert: RootCertStore::empty(),
+        }
+    }
+
+    /// Build a TLS client with parameters provided to builder:
+    ///
+    /// If all CA certificate, Client Certificate and Client key are provided,
+    /// it builds a TLS client with mutual authentication
+    ///
+    /// If none of certificates and key are *not* provided, it builds a TLS
+    /// client with *root* CA certificate
+    ///
+    /// Otherwise it returns an error
+    pub fn build(self) -> RedisResult<Client> {
+        let tls_params = {
+            match (self.client_key, self.client_certs, self.root_cert) {
+                (Some(client_key), client_certs, root_cert) => Some(TlsConnParams {
+                    client_key,
+                    client_certs,
+                    root_cert,
+                }),
+                (None, client_certs, root_cert)
+                    if client_certs.is_empty() && root_cert.is_empty() =>
+                {
+                    None
+                }
+                _ => fail!((
+                    ErrorKind::InvalidClientConfig,
+                    "TLS client config not ready"
+                )),
+            }
+        };
+
+        Ok(Client {
+            connection_info: ConnectionInfo {
+                addr: ConnectionAddr::TcpTls {
+                    host: self.host,
+                    port: self.port,
+                    insecure: false,
+                    tls_params,
+                },
+                redis: RedisConnectionInfo::default(),
+            },
+        })
+    }
+
+    /// Loads a client certificate in PEM format as a byte stream
+    /// In case of errors loading the certificate it returns an error
+    pub fn client_certs(&mut self, mut reader: &mut dyn BufRead) -> RedisResult<()> {
+        self.client_certs =
+            certs(&mut reader).map(|mut certs| certs.drain(..).map(Certificate).collect())?;
+
+        Ok(())
+    }
+
+    /// Loads a client key in PEM format as a byte stream
+    /// In case of errors loading the key it returns an error
+    pub fn client_key(&mut self, reader: &mut dyn BufRead) -> RedisResult<()> {
+        self.client_key = Some(load_key(reader)?);
+        Ok(())
+    }
+
+    /// Loads a CA root certificate in PEM format as a byte stream
+    /// In case of errors loading the certificate it returns an error
```

Comment:
doc nit:

```suggestion
    /// Loads a CA root certificate in PEM format as a byte stream.
    ///
    /// In case of errors loading the certificate it returns an error.
```